freqtrade/freqai/data_kitchen.py

```diff
@@ -1126,13 +1126,57 @@ def set_all_pairs(self) -> None:
             if pair not in self.all_pairs:
                 self.all_pairs.append(pair)
 
+    def extract_corr_pair_columns_from_populated_indicators(
+        self,
+        dataframe: DataFrame
+    ) -> Dict[str, DataFrame]:
+        """
+        Find the columns of the dataframe corresponding to the corr_pairlist, save them
+        in a dictionary to be reused and attached to other pairs.
+        :params:
+        :dataframe: fully populated dataframe (current pair + corr_pairs)
+        :return:
+        :corr_dataframes: dictionary of dataframes to be attached to other pairs in same candle.
+        """
+        corr_dataframes: Dict[str, DataFrame] = {}
+        pairs = self.freqai_config["feature_parameters"].get("include_corr_pairlist", [])
+
+        for pair in pairs:
+            coin = pair.split('/')[0]
+            pair_cols = [col for col in dataframe.columns if coin in col]
+            corr_dataframes[pair] = dataframe.filter(pair_cols, axis=1)
+
+        return corr_dataframes
+
+    def attach_corr_pair_columns(self, dataframe: DataFrame,
+                                 corr_dataframes: Dict[str, DataFrame],
+                                 current_pair: str) -> DataFrame:
+        """
+        Attach the existing corr_pair dataframes to the current pair dataframe before training
+        :params:
+        :dataframe: current pair strategy dataframe, indicators populated already
+        :corr_dataframes: dictionary of saved dataframes from earlier in the same candle
+        :current_pair: current pair to which we will attach corr pair dataframe
+        :return:
+        :dataframe: current pair dataframe of populated indicators, concatenated with corr_pairs
+                    ready for training
+        """
+        pairs = self.freqai_config["feature_parameters"].get("include_corr_pairlist", [])
+
+        for pair in pairs:
+            if current_pair not in pair:
+                dataframe = pd.concat([dataframe, corr_dataframes[pair]], axis=1)
```

Comment:
Can you explain this logic?
`pairs` is a list of pairs (e.g. `['BTC/USDT', 'XRP/USDT', ...]`)

do we assume `current_pair` to be `BTC/USDT`? (then why not use == for comparison?)

no matter what the intend here is (or the content of current pair) - using `in` for pair comparisons is quite certainly wrong.
Crypto pairs are often named similarily - you could have `BTC/USD` and `BTC/USDT` - the first is "in" the latter - but that's not the comparison you're looking for.

There's also a pair `T/USDT` - so every coin that ends in `T` would have the condition "true" - which would most certainly have unintended side-effects.

Reply:
the `if current_pair not in pair:` is a check to ensure that if a pair exists simultaneously on the `whitelist` and the `corr_pairlist`, we will not add the same indicators to itself. That is all.

For the point about `T/USDT`, it is a fair point. And the only way around this would be to switch FreqAI away from using `coin = pair.split('/')[0]` in `populate_any_indicators` to simply using the full `pair` string. 

But - that will be a massive breaking change, everyone will need to update their `populate_any_indicator` functions, presumably with the following edit:

```python
coin = pair #.split('/')[0]  <-- removing the split call *should* be enough.  
```

Although small - it still would require everyone to do it.

Reply:
`T/USDT` is an extreme sample - i've never heared of it before - and probably will never again unless my bot happens to buy it - but it's as good as any to serve as an example.
ETH/USDT and xETH/USDT (Wrapped ETH iirc) is a similar example - and many other derivatives which have (intentionally or not) a similar name).

in hindsight it's a bit unfortunate that we use `f"%-{coin}rsi-period_{t}"` - not `f"%-{coin}-rsi-period_{t}"` - as that would give us a clear indication as to where the coin name is "finished" and where the indicator name starts.


with indicator naming - we do take a LOT of faith in users to keep the same nomenclature (lowercase, same identical syntax, ...) - which is in the samples - but not documented - and also not really a requirement ...

let's take a diff. example of a problematic "in" condition:
we use  `informative[f"{coin}bb_lowerband-period_{t}"] = bollinger["lower"]` ... but maybe the user deletes (and recreates) this and it becomes `informative[f"{coin}bb_LOWERBAND-period_{t}"] = bollinger["lower"]` (uppercased lowerband). (obvoiusly this could happen with any coin/indicator - as there's too many coins to count - with most possible 3/4/5 letter symbols).

Now we'd have a match on `BAND/USDT` (where the coin is `BAND` - which is a valid, listed coin on binance).
The naming convention only specifies the starting of a column (`%- / `%%-` / `&-`) - not the rest of the name.

It's not a problem as long as the "naming" remains what it is - just a column name.
the moment we attempt to put bigger meaning into the name (like figuring out the pair name) - we might encounter problems - where the naming will no longer be reliable.

---

While I'm not sure if it'll fix all problems, maybe using `%-{coin}` (and all other documented counterparts) as search-term for columns could be an option? so the column has to start with one of these (3?) signs to be identified as "belongs to this coin".

It'll still not fix all problems (`T` / `TRX` -> `%-Tmfi`, `%-TRXmfi`) - but will be slightly safer than `T in column_name` ...

Reply:
This is now handled inthe following way:

Users are encouraged via the example strat and docs, to use the full pair string when making their features in `populate_any_indicators()`. Our backend looks for the full pair string in one of the following forms:

```python
valid_strs = [f"%-{pair}", f"%{pair}", f"%_{pair}"]
```

If none of these are matched, but the user has pairs in their `include_corr_pairlist`, then FreqAI will log a warning:

```python
                logger.warning("Couldn't cache corr_pair dataframes for improved performance. "
                               "Consider ensuring that the full coin/stake, e.g. XYZ/USD, "
                               "is included in the column names when you are creating features "
                               "in `populate_any_indicators()`.")
```

And then revert to old behavior.

Reply:
This is now handled inthe following way:

Users are encouraged via the example strat and docs, to use the full pair string when making their features in `populate_any_indicators()`. Our backend looks for the full pair string in one of the following forms:

```python
valid_strs = [f"%-{pair}", f"%{pair}", f"%_{pair}"]
```

If none of these are matched, but the user has pairs in their `include_corr_pairlist`, then FreqAI will log a warning:

```python
                logger.warning("Couldn't cache corr_pair dataframes for improved performance. "
                               "Consider ensuring that the full coin/stake, e.g. XYZ/USD, "
                               "is included in the column names when you are creating features "
                               "in `populate_any_indicators()`.")
```

And then revert to old behavior.